pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/relocation/SegmentRelocator.java

```diff
@@ -91,10 +97,66 @@ public SegmentRelocator(PinotHelixResourceManager pinotHelixResourceManager,
         Math.min(taskIntervalInMs, config.getSegmentRelocatorExternalViewCheckIntervalInMs());
     _externalViewStabilizationTimeoutInMs =
         Math.min(taskIntervalInMs, config.getSegmentRelocatorExternalViewStabilizationTimeoutInMs());
+    if (config.isSegmentRelocatorRebalanceTablesSequentially()) {
+      _waitingTables = ConcurrentHashMap.newKeySet();
+      _waitingQueue = new LinkedBlockingQueue<>();
+      _executorService.submit(() -> {
+        LOGGER.info("Rebalance tables sequentially");
+        while (!Thread.currentThread().isInterrupted()) {
+          rebalanceWaitingTable(this::rebalanceTable);
+        }
+        return null;
+      });
+    } else {
+      _waitingTables = null;
+      _waitingQueue = null;
+    }
   }
 
   @Override
   protected void processTable(String tableNameWithType) {
+    if (_waitingTables == null) {
+      LOGGER.debug("Rebalance table: {} immediately", tableNameWithType);
+      _executorService.submit(() -> rebalanceTable(tableNameWithType));
+      return;
+    }
+    putTableToWait(tableNameWithType);
+  }
+
+  @VisibleForTesting
+  synchronized void putTableToWait(String tableNameWithType) {
```

Comment:
Since we are using a concurrent set, and there is no capacity limit for the waiting queuem we can remove the `synchronized` and do:
```
  if (_waitingTables.add(tableNameWithType)) {
    _waitingQueue.offer(tableNameWithType);
  } else {
    LOGGER.debug(already exist...);
  }
```

Reply:
neat